mail/protocols/smtp/src/main/java/com/fsck/k9/mail/transport/smtp/SmtpTransport.kt

```diff
@@ -551,24 +556,57 @@ class SmtpTransport(
     private fun saslOAuth(method: OAuthMethod) {
         Log.d("saslOAuth() called with: method = $method")
         retryOAuthWithNewToken = true
-
-        val primaryEmail = oauthTokenProvider?.primaryEmail
-        val primaryUsername = primaryEmail ?: username
-
-        try {
-            attempOAuth(method, primaryUsername)
-        } catch (negativeResponse: NegativeSmtpReplyException) {
-            Log.w(negativeResponse, "saslOAuth: failed to authenticate.")
-            if (negativeResponse.replyCode != SMTP_AUTHENTICATION_FAILURE_ERROR_CODE) {
-                throw negativeResponse
+        checkNotNull(oauthTokenProvider) { "No OAuth2TokenProvider available." }
+
+        val users = buildSet {
+            // add the given username to the list of users
+            add(username)
+            // add all the usernames we can fetch from the id_token to be used in case from
+            // the given username is fails to authenticate.
```

Comment:
```suggestion
            // add all the usernames we can fetch from the id_token to be used in case
            // the given username fails to authenticate.
```